### PR TITLE
UCM/CONFIGURE: UCM_MODULE_LDFLAGS setting per OSes

### PR DESCRIPTION
**Pull request by hiroyuki-sato**
## What

This PR doesn't change the current behavior.
UCM_MODULE_LDFLAGS setting per OSes

## Why ?

For porting UCX to macOS, we need to set `UCM_MODULE_LDFLAGS` option. 
It doesn't the same setting as the Linux platform.

## How ?

Switch that setting using `case` statement.
I'll add macOS setting later.

### CODE REVIEW

**Comment by shamisp** [comment]
ok to test

**Comment by hiroyuki-sato** [comment]
```
##[error]The job running on agent swx-rdmz-ucx-gpu-01-01 ran longer than the maximum time of 300 minutes. For more information, see https://go.microsoft.com/fwlink/?linkid=2077134
```

Could anyone rerun AZP?

**Comment by shamisp** [comment]
The infrastructure makes sense.

**Review by shamisp — APPROVED** [review]

File: src/ucm/configure.m4 [hunk 1]
```diff
@@ -4,8 +4,13 @@
...
 
-AC_SUBST([UCM_MODULE_LDFLAGS],
-         ["-Xlinker -z -Xlinker interpose -Xlinker --no-as-needed"])
+AS_CASE([$host_os],
+    [linux*], [
+        AC_SUBST([UCM_MODULE_LDFLAGS],
+                 ["-Xlinker -z -Xlinker interpose -Xlinker --no-as-needed"])
+    ],
+    [AC_MSG_ERROR([UCM_MODULE_LDFLAGS=PORT ME])]
+)
```
Comment (yosefe) [inline src/ucm/configure.m4:13]: need to test for linker flags support, potentially by AC_LINK_IFELSE

Comment (hiroyuki-sato) [reply src/ucm/configure.m4]: I added AC_LINK_IFELSE test.

**Comment by hiroyuki-sato** [comment]
Hello, @shamisp and @yosefe Thank you for your comments.

I'm not familiar with the Autoconf. Please let me know if there's a better way.

```
configure:25465: gcc -o conftest    -Xlinker -z -Xlinker interpose -Xlinker --no-as-needed conftest.c -lpthread -lrt -lrt  -ldl >&5
configure:25465: $? = 0
```

File: src/ucm/configure.m4 [hunk 2]
```diff
@@ -4,8 +4,24 @@
 # See file LICENSE for terms.
 #
 
-AC_SUBST([UCM_MODULE_LDFLAGS],
-         ["-Xlinker -z -Xlinker interpose -Xlinker --no-as-needed"])
+AS_CASE([$host_os],
+    [linux*], [
```
Comment (yosefe) [inline src/ucm/configure.m4:8]: no need to check host os. if AC_LINK_IFELSE -s ok, regardless of OS, we can use those flags

File: src/ucm/configure.m4 [hunk 3]
```diff
@@ -4,8 +4,24 @@
...
+        SAVE_LDFLAGS="$LDFLAGS"
+
+        LDFLAGS="$LDFLAGS -Xlinker -z -Xlinker interpose -Xlinker --no-as-needed"
+
+        AC_LINK_IFELSE([AC_LANG_PROGRAM([[#include <stdio.h>]])],[
```
Comment (yosefe) [inline src/ucm/configure.m4:13]: the program should be `[[ int main(void) { return 0; } ]]`

File: src/ucm/configure.m4 [hunk 4]
```diff
@@ -4,8 +4,24 @@
...
+            AC_SUBST([UCM_MODULE_LDFLAGS],
+                     ["-Xlinker -z -Xlinker interpose -Xlinker --no-as-needed"])
+
+        ],[
+            AC_MSG_ERROR([Required -Xlinker option does not work])
+        ])
+
+        LDFLAGS="$SAVE_LDFLAGS"
+    ],
+    [AC_MSG_ERROR([UCM_MODULE_LDFLAGS=PORT ME])]
```
Comment (yosefe) [inline src/ucm/configure.m4:23]: `[AC_MSG_ERROR([UCM linker flags are not supported])]`

**Comment by hiroyuki-sato** [comment]
Hello, @yosefe. Thank you for your comment.

Could you tell me about your comment?
Does your proposed way is the below, correct?

IMO, each OS use different linker option. Anyway, I'll try to implement it.

```bash
#!/bin/sh

# 1st option is for linux
LINKER_FLAG_CANDIDATES="
-Xlinker -z -Xlinker interpose -Xlinker --no-as-needed
this linker option is for darwin
this linker option is for freebsd
"

SAVE_IFS="$IFS"

IFS="
"

UCM_MODULE_LDFLAGS=""

#
# AC_LINK_IFELSE part.
#  build-ok: set UCM_MODULE_LDFLAGS
#  build-ng: try next option.
#
for opts in $LINKER_FLAG_CANDIDATES ; do

  #
  # compile test code using liner option.
  # It returns 0 when build pass
  #
  gcc -o test $opts
  if [ $? -eq 0 ] ; then
     UCM_MODULE_LDFLAGS=$opts
     break;
  fi
done

if [ -z "$UCM_MODULE_LDFLAGS" ] ; then
  echo "UCM linker flags are not supported"
  exit 1
fi

IFS="$SAVE_IFS"

echo $UCM_MODULE_LDFLAGS
```

**Comment by hiroyuki-sato** [comment]
Hello, @yosefe . Please take a look when you get a chance.

I don't use `else` part in `AS_IF` macros. Because It will be a very complicated code.

## 1. For macOS porting

I'll add the following part later

```m4
#
# darwin
#
AS_IF([test "x$UCM_MODULE_LDFLAGS" = "x"],[
    UCM_MODULE_LDFLAGS_TEST="-Wl,Darwin_specific_linker_option "
    LDFLAGS="$SAVE_LDFLAGS $UCM_MODULE_LDFLAGS_TEST"
    AC_LINK_IFELSE([AC_LANG_PROGRAM([])],[
        AC_SUBST([UCM_MODULE_LDFLAGS],[$UCM_MODULE_LDFLAGS_TEST])
    ],[])
],[])
```

## 2. Failed implemenetation

### 2.1 Using for loop

First I tried the following codes.

```m4
UCM_MODULE_LDFLAGS_CANDIDATES="
-Xlinker -z -Xlinker interpose -Xlinker --no-as-needed
"

SAVE_IFS="$IFS"
SAVE_LDFLAGS="$LDFLAGS"
IFS="
"

for linker_opts in $UCM_MODULE_LDFLAGS_CANDIDATES ; do
    LDFLAGS="$LDFLAGS $linker_opts"
    AC_LINK_IFELSE([AC_LANG_PROGRAM([[]])],[
        AC_SUBST([UCM_MODULE_LDFLAGS],[$linker_opts])
        break
    ],[])
done
AS_IF([test "x$UCM_MODULE_LDFLAGS" = "x"],[
    AC_MSG_ERROR([UCM linker flags are not supported])
],[])

IFS="$SAVE_IFS"
LDFLAGS="$SAVE_LDFLAGS"
```

This version outputs the following error.

```
./configure: line 2250: printf %s\n: command not found
./configure: line 2258: printf %s\n: command not found
./configure: line 2268: printf %s\n: command not found
./configure: line 421: printf %s\n: command not found
./configure: line 423: printf %s\n: command not found
./configure: line 8: printf %s\n: command not found
./configure: line 44: printf %s\n: command not found
./configure: line 52: printf %s\n: command not found
./configure: line 52: printf %s\n: command not found
```

## 2.2 without `AC_LANG_PROGRAM([[]])]`

```m4
AC_LINK_IFELSE([[[ int main(void) { return 0; } ]]],[
    AC_SUBST([UCM_MODULE_LDFLAGS],[$UCM_MODULE_LDFLAGS_TEST])
],[])
```

```
src/ucm/configure.m4:15: warning: AC_LANG_CONFTEST: no AC_LANG_SOURCE call detected in body
../../lib/autoconf/lang.m4:193: AC_LANG_CONFTEST is expanded from...
../../lib/autoconf/general.m4:2661: _AC_LINK_IFELSE is expanded from...
```

**Review by yosefe — APPROVED** [review]

**Pull request merged**